src/main/java/com/artipie/asto/Key.java

```diff
@@ -178,6 +192,11 @@ public String string() {
             return parent;
         }
 
+        @Override
+        public List<String> parts() {
+            return this.parts.stream().collect(Collectors.toList());
```

Comment:
You can just use `return Collections.unmodifiableList(this.parts)` without creating a stream here.